File: FullDeviceInfo/Sources/FullDeviceInfoObjc/SaveKeychain.m
//
//  SaveKeychain.m
//  FullDeviceInfo
//
//  Created by pastry on 27/08/25.
//

#import "SaveKeychain.h"
#import <Security/Security.h>


#define kService @"com.fullDeviceInfo.keychain"
#define kAccount @"com.fullDeviceInfo.account"

@implementation SaveKeychain


+ (instancetype)sharedInstance {
    static SaveKeychain *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        
    }
    return self;
}


- (BOOL)saveData:(NSData *)data
          forKey:(NSString *)key {
    
    if (!data || !key) {
        return NO;
    }
    
    NSMutableDictionary *query = [NSMutableDictionary dictionary];
    [query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    [query setObject:kService forKey:(__bridge id)kSecAttrService];
    [query setObject:kAccount forKey:(__bridge id)kSecAttrAccount];
    [query setObject:key forKey:(__bridge id)kSecAttrGeneric];
    
    SecItemDelete((__bridge CFDictionaryRef)query);
    
    [query setObject:data forKey:(__bridge id)kSecValueData];
    
    OSStatus status = SecItemAdd((__bridge CFDictionaryRef)query, NULL);
    
    return (status == errSecSuccess);
}

- (nullable NSData *)loadDataForKey:(NSString *)key {
    
    if (!key) {
        return nil;
    }
    
    NSMutableDictionary *query = [NSMutableDictionary dictionary];
    [query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    [query setObject:kService forKey:(__bridge id)kSecAttrService];
    [query setObject:kAccount forKey:(__bridge id)kSecAttrAccount];
    [query setObject:key forKey:(__bridge id)kSecAttrGeneric];
    [query setObject:(__bridge id)kSecMatchLimitOne forKey:(__bridge id)kSecMatchLimit];
    [query setObject:(id)kCFBooleanTrue forKey:(__bridge id)kSecReturnData];
    
    CFTypeRef result = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &result);
    
    if (status == errSecSuccess && result != NULL) {
        NSData *data = (__bridge_transfer NSData *)result;
        return data;
    }
    
    return nil;
}

- (BOOL)deleteDataForKey:(NSString *)key {
    
    if (!key) {
        return NO;
    }
    
    NSMutableDictionary *query = [NSMutableDictionary dictionary];
    [query setObject:(__bridge id)kSecClassGenericPassword forKey:(__bridge id)kSecClass];
    [query setObject:kService forKey:(__bridge id)kSecAttrService];
    [query setObject:kAccount forKey:(__bridge id)kSecAttrAccount];
    [query setObject:key forKey:(__bridge id)kSecAttrGeneric];
    
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
    
    return (status == errSecSuccess || status == errSecItemNotFound);
}


- (BOOL)saveString:(NSString *)string
            forKey:(NSString *)key {
    
    if (!string || !key) {
        return NO;
    }
    
    NSData *data = [string dataUsingEncoding:NSUTF8StringEncoding];
    return [self saveData:data forKey:key];
}

- (nullable NSString *)loadStringForKey:(NSString *)key {
    
    NSData *data = [self loadDataForKey:key];
    if (data) {
        return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    }
    return nil;
}

@end

File: FullDeviceInfo/Sources/FullDeviceInfo/SaveKeychain.swift
//
//  SaveKeychain.swift
//
//
//  Created by pastry on 27/08/25.
//

import UIKit
import Security

private let service = "com.fullDeviceInfo.keychain"

class SaveKeychain: NSObject {
    
    @MainActor static let shared = SaveKeychain()
    
    
    @discardableResult
    func save(string: String, forKey key: String) -> Bool {
        guard let data = string.data(using: .utf8) else { return false }
        return save(data: data, forKey: key)
    }
    
    func loadString(forKey key: String) -> String {
        guard let data = loadData(forKey: key) else { return "" }
        return String(data: data, encoding: .utf8) ?? ""
    }
    
    
    
    
    
    
    
    
    
    
    @discardableResult
    private func save(data: Data, forKey key: String) -> Bool {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        SecItemDelete(query as CFDictionary)
        
        let status = SecItemAdd(query as CFDictionary, nil)
        
        return status == errSecSuccess
    }
    
    private func loadData(forKey key: String) -> Data? {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        return (result as? Data)
    }
    
    @discardableResult
    func update(data: Data, forKey key: String) -> Bool {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
        
        let attributes: [String: Any] = [
            kSecValueData as String: data
        ]
        
        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        
        return status == errSecSuccess
    }
    
    @discardableResult
    func update(string: String, forKey key: String) -> Bool {
        guard let data = string.data(using: .utf8) else { return false }
        return update(data: data, forKey: key)
    }
    
    @discardableResult
    func delete(forKey key: String) -> Bool {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        
        return status == errSecSuccess
    }
    
    func exists(forKey key: String) -> Bool {
        return loadData(forKey: key) != nil
    }
    
    @discardableResult
    func clearAll() -> Bool {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        
        return status == errSecSuccess
    }
    
    func getAllKeys() -> [String] {
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecReturnAttributes as String: true,
            kSecMatchLimit as String: kSecMatchLimitAll
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess,
              let items = result as? [[String: Any]] else {
            return []
        }
        
        return items.compactMap { $0[kSecAttrAccount as String] as? String }
    }
}
